### ISSUE TITLE
Different indels with `--mark-del '-'`

### ISSUE THREAD

**Issue by DrYak** [samtools/bcftools#1963]
While testing a new chain-file parser for our frameshifting_deletions_checks tool, I've noticed that bcftools consensus does not produce the same indels whether `--mark-del '-'` is set or not.

When `--mark-del '-'` is set, it calls a consensus with 6bp-long deletion.
When no `--mark-del` is set, it calls a consensus with 7bp-long gap(deletion) in the consensus and 1bp-long gap (insertion) in the reference.

Comparison of the chain files:

```diff
--- a/tests/test_frameshift_deletions_checks/stopgain_with_insertions.chain
+++ b/tests/test_frameshift_deletions_checks/stopgain_with_insertions_nogap.chain
@@ -1,5 +1,8 @@
-chain 29903 NC_045512.2 29903 + 0 29903 NC_045512.2 29905 + 0 29905 1
-7889 0 1
-21415 0 1
+chain 29893 NC_045512.2 29903 + 0 29903 NC_045512.2 29896 + 0 29896 1
+4906 1 0
+2982 0 1
+2417 2 0
+13826 7 1
+5163 0 1
 599
 ```
 
Comparison of (a subset) of the consensus:
```text
24130     24140     24150
    |    .    |    .    |    .
tttaacggcc-------ttttgccaccttt
                --
tttaacggcct------tttgccacctttg
```

I will upload the files necessary to replicate in a few minutes...

**Comment by DrYak** [comment]
[bcftools_issue_1963.tar.gz](https://github.com/samtools/bcftools/files/12071822/bcftools_issue_1963.tar.gz) : Everything except the alignments, due to GitHub size limitations.
[Polybox (NextCloud) link](https://polybox.ethz.ch/index.php/s/IBopy2kzxI4hEjW): Folder with everything, including alignments.

I'll add explanation to replicate in a few minutes

**Comment by DrYak** [comment]
The bcftools consensus outputs:
- `stopgain_with_insertions.fasta`  and `stopgain_with_insertions.chain`
  - output files that we normally generate from our pipeline (V-pipe) with deletion marked with `-` in the consensus.
- `stopgain_with_insertions_nogap.fasta`  and `stopgain_with_insertions_nogap.chain`
  - generated in a similar fashion, but without the `--mark '-'` options (deletions are instead reported in the chain file).

The files:
- `REF_aln_trim.bam` (and `...bai`) the alignment from which the above consensuses were generated (literaly the same file).
- `stopgain_with_insertions.cram`  (and `...crai`)  same alignment, but much compressed as .cram instead for space saving (we use that in the unit tests).
  - you can use either, they have the same content.
- `NC_045512.2.fasta` (and `....fai`) the reference genome (SARS-CoV-2, original WuHan strain) against which the reads are aligned.
- `coverage.tsv.gz` coverage per position (1-based)
- `temp.consensus.bcftools.bcf.gz` output of the pile-up chain
- `consensus.bcftools.bcf.gz` (and `...csi`) same above, after filtering by [V-pipe's enhance_bcf.py](https://github.com/cbg-ethz/V-pipe/blob/master/workflow/scripts/enhance_bcf.py)
   - (it's used for calling ambiguous bases, SHOULD NOT impact our issue here)
- `coverage_mask_lowcoverage.bed` region with a too low coverage that should be masked as `n` in the final output.

Processus:
- `mkconsensus` script that runs the consensus calling
  - it's based on how [V-pipe's rule consensus_bcftools](https://github.com/cbg-ethz/V-pipe/blob/master/workflow/rules/consensus.smk#L8) does it.
- [Dependencies](https://github.com/cbg-ethz/V-pipe/blob/master/workflow/envs/bcftools.yaml):
dependencies:
  - bcftools = 1.13
  - cyvcf2 = 0.30.11

**Comment by pd3** [comment]
Thank you for the thorough test case. However, it's a bit too thorough, you've included an entire pipeline :)
Any chance you could narrow the issue down to the first step where you're seeing the difference that bothers you? From a cursory glance it is not clear which of the 6+ commands, that include a custom python script, is causing the difference.

**Comment by DrYak** [comment]
Sorry, yes, I wanted to make sure everything was there.

The difference arises in the last step:

```bash
bcftools consensus \
        --output ${output_fname_fasta} \
        --chain ${output_fname_chain} \
        --fasta-ref NC_045512.2.fasta \
        --mark-del - \
        --mask coverage_mask_lowcoverage.bed --mask-with n \
        -H A \
        -i "INFO/AD[0]<INFO/AD[*]" \
        consensus.bcftools.bcf.gz
```

If I remove the `--mark-del -`  part, the deletion (as described by the chain file) are not the same and off by 1 base.

**Comment by pd3** [comment]
Right, okay. 

I believe the program is doing the correct thing: the chain file constitutes a mapping between coordinate systems of two sequences, the original fasta file and the output fasta file. Importantly, these sequences are strings of characters, not true biological sequences, therefore as far as the chain files are concerned, dashes are handled the same way as A,C,G,T.  Since in one case the deletions are actually deleted from the sequence (`--mark-del -` option not given) and in the other the deleted sequence is only replaced with dashes (`--mark-del -` option given), the coordinates change and one must expect different chain files on output.

In other words, as far as chain files are concerned, `--mark-del -` behaves as if deletions were not applied at all.

**Comment by DrYak** [comment]
My expectations were that the deletions should be exactly the same:
 - when `--mard-del -` is given: the deletion is marked with `-`
 - when `--mard-del -` is not given: the exact same deletion (which was marked above by `-`) should now instead be indicated in the .chain file.
 
 What I get instead:
 - when using `--mark-del -` a deletion with 6 marks (`------`) at a position.
 - when not using `--mark-del` a deletion of lenght 7, plus one insertion at the same position.
 
 I am wondering if it's me who's reading the .chain wrong.

**Comment by pd3** [comment]
What I was trying to explain in the comment https://github.com/samtools/bcftools/issues/1963#issuecomment-1692909074 above is that the chain files work with strings and map matching positions between two fasta files. If you replace a deleted sequence with dashes `---`, the coordinate system is not changed by it, hence the chain file will not report it.

**Comment by pd3** [comment]
I believe this can be marked as resolved, please complain if not.
